### ISSUE TITLE
Any multi-threading examples?

### ISSUE THREAD

**Issue by geovex** [MoAlyousef/fltk-rs#31]
Is it even possible?

**Comment by MoAlyousef** [comment]
Hi geovex

fltk supports multithreading. This is a spin-off of the hello_button example using threading:
```rust
use fltk::{app::*, button::*, frame::*, window::*};
use std::{thread, time};

fn main() {
    let app = App::default();
    let mut wind = Window::default()
        .with_size(400, 300)
        .center_screen()
        .with_label("Hello");
    let mut frame = Frame::new(0, 0, 400, 200, "");
    let mut but = Button::new(160, 210, 80, 40, "Click me!");
    but.set_callback(Box::new(|| {
        thread::spawn(move || {
            thread::sleep(time::Duration::from_secs(5));
            frame.set_label("Hello World");
        });
    }));
    wind.show();
    app.run().unwrap();
}
```

**Comment by geovex** [comment]
fltk doc [mention](https://www.fltk.org/doc-1.3/advanced.html) `FL::lock` or  `FL::awake`. Shouldn't we use them in child threads?

**Comment by MoAlyousef** [comment]
The App struct when run is called, calls Fl::lock:
```rust
    pub fn run(&self) -> Result<(), FltkError> {
        lock()?;
        return run();
    }
```
This enables the internal lock (according to the fltk documentation), which is then used internally by widgets for any method mutating the widget.

**Comment by geovex** [comment]
> A worker thread, when it wants to alter the value of a widget, can acquire the lock using Fl::lock(), update the widget, then release the lock using Fl::unlock(). Acquiring the lock ensures that the worker thread can update the widget, without any risk that the main() thread will attempt to redraw the widget whilst it is being updated.

fltk_rs does not have that. Also I didn't find internal locking in fltk methods, only in main_loop implementations (but it's possible i didn't look where I should).

**Comment by MoAlyousef** [comment]
It's done in the C++ side using a macro:
```c++
#define LOCK(x)                                                                \
  Fl::lock();                                                                  \
  x;                                                                           \
  Fl::unlock();                                                                \
  Fl::awake();
```

**Comment by geovex** [comment]
```rust
extern crate fltk;

use fltk::{app::*, button::*, frame::*, window::*};
use std::{thread, time};

fn main() {
    let app = App::default();
    let mut wind = Window::default()
        .with_size(400, 300)
        .center_screen()
        .with_label("Hello");
    let frame = Frame::new(0, 0, 400, 200, "");
    let mut but = Button::new(160, 210, 80, 40, "Click me!");
    but.set_callback(Box::new(|| {
        let mut frame = frame.clone();
        thread::spawn(move|| {
            for i in 0..1000 {
                thread::sleep(time::Duration::from_millis(10));
                frame.set_label(format!("Hello {}", i).as_str());
            }
        });
    }));
    wind.show();
    app.run().unwrap();
}
```

**Comment by MoAlyousef** [comment]
I honestly can't reproduce the crash. What OS are you using? Also what version or branch are you using?

**Comment by geovex** [comment]
I used archlinux
Fltk_rs 0.2.8 from crates.io

**Comment by MoAlyousef** [comment]
Can you try updating to 0.2.9? The first addition ```.define("OPTION_USE_THREADS", "ON")``` was quite recent, and the locking as well on the C++ side. It was delayed on a local copy of the repo because of some other issues. 
If it's fixed with 0.2.9 I'll yank 0.2.8.
Thank you

**Comment by geovex** [comment]
I honestly think we only need to implement awake with callback somehow to have some kind of thread-safety and type soundness.

**Comment by MoAlyousef** [comment]
Fltk (the C++ library) does offer a mechanism for that, I'll see what I can do to implement it in Rust.

**Comment by geovex** [comment]
Ok... version from `master` works (looks like it). I will try to use it in heavily threaded project as a test.

**Comment by MoAlyousef** [comment]
I was able to reproduce the crash on 0.2.8 and I discovered I had forgetten to publish 0.2.9. So that's done now. I also no longer get a crash on 0.2.9. I'll wait for your testing as well.
Thanks again

**Comment by MoAlyousef** [comment]
I'll be closing this issue since it isn't too specific and appears to be fixed in 0.2.9 and later. I prefer if other issues appear for them to have their own issue tracker.
